src/clj/puppetlabs/services/request_handler/request_handler_core.clj

```diff
@@ -173,40 +176,68 @@
                cert-count
                " found"))))))
 
-(defn auth-maybe-with-client-header-info
-  "Return authentication info based on client headers"
-  [config headers]
-  (let [header-dn-name   (:ssl-client-header config)
-        header-dn-val    (get headers header-dn-name)
-        header-auth-name (:ssl-client-verify-header config)
-        header-auth-val  (get headers header-auth-name)
-        header-cert-val  (get headers header-client-cert-name)]
-    (if (:allow-header-cert-info config)
-      (-> (header-auth-info header-dn-name
-                            header-dn-val
-                            header-auth-val)
-          (assoc :client-cert (header-cert header-cert-val)))
-      (do
-        (doseq [[header-name header-val] {header-dn-name           header-dn-val
-                                          header-auth-name         header-auth-val
-                                          header-client-cert-name  header-cert-val}]
-          (if header-val
+(defn ssl-auth-info
+  "Get map of client authentication info from the supplied
+   `java.security.cert.X509Certificate` object.  If the supplied object is nil,
+   the information returned would represent an 'unauthenticated' client."
+  [ssl-client-cert]
+  (if ssl-client-cert
+    (let [cn (ssl-utils/get-cn-from-x509-certificate ssl-client-cert)
+          authenticated (not (empty? cn))]
+      (log/debugf "CN '%s' provided by SSL certificate.  Authenticated: %s."
+                  cn authenticated)
+      {:client-cert-cn cn
+       :authenticated  authenticated})
+    (do
+      (log/debugf "No SSL client certificate provided. "
+                  "Treating client as unauthenticated.")
+      unauthenticated-client-info)))
+
+(defn client-auth-info
+  "Get map of client authentication info for the client.  Map has the following
+  keys:
+
+  * :client-cert - A `java.security.cert.X509Certificate` object or nil
+  * :client-cert-cn - The CN (Common Name) of the client, typically associated
+                      with the CN attribute from the Distinguished Name
+                      in an X.509 certificate's Subject.
+  * :authenticated - A boolean representing whether or not the client is
+                     considered to have been successfully authenticated.
+
+  Parameters:
+
+  * config - Map of configuration data
+  * request - Ring request containing client data"
+  [config request]
+  (if-let [authorization (:authorization request)]
+    {:client-cert    (ring-auth/authorized-certificate request)
+     :client-cert-cn (ring-auth/authorized-name request)
+     :authenticated  (true? (ring-auth/authorized-authentic? request))}
+    (let [headers (:headers request)
```

Comment:
ring actually had a `get-header` function which allows you to grab headers by name directly from the request object, but this is totally fine with me as it is.